storage/innobase/handler/ha_innodb.cc

```diff
@@ -9023,91 +9022,100 @@ ha_innobase::index_read(
 			m_prebuilt->srch_key_val_len,
 			index,
 			(byte*) key_ptr,
-			(ulint) key_len);
+			key_len);
 
 		DBUG_ASSERT(m_prebuilt->search_tuple->n_fields > 0);
 	} else {
+		ut_ad(find_flag != HA_READ_KEY_EXACT);
 		/* We position the cursor to the last or the first entry
 		in the index */
 
 		dtuple_set_n_fields(m_prebuilt->search_tuple, 0);
 	}
 
-	page_cur_mode_t	mode = convert_search_mode_to_innobase(find_flag);
-
-	ulint	match_mode = 0;
-
-	if (find_flag == HA_READ_KEY_EXACT) {
-
-		match_mode = ROW_SEL_EXACT;
-
-	} else if (find_flag == HA_READ_PREFIX_LAST) {
+	page_cur_mode_t	mode = PAGE_CUR_LE;
+	m_last_match_mode = 0;
 
-		match_mode = ROW_SEL_EXACT_PREFIX;
+	switch (find_flag) {
+	case HA_READ_KEY_EXACT:
```

Comment:
I understand this is for performance but it is leading to code duplication with convert_search_mode_to_innobase().

1. Perhaps inlining the convert_search_mode_to_innobase() could have similar effect ?
2. The function could return a std::pair of cursor_mode and match_mode and match_mode can be ignored in other callers.

Reply:
I see that `convert_search_mode_to_innobase()` was missing `static`. I am going to implement it as follows, to avoid any run-time overhead:
```c++
/** Convert a MariaDB search mode to an InnoDB search mode.
@tparam last_match whether last_match_mode is to be set
@param find_flag       MariaDB search mode
@param mode            InnoDB search mode
@param last_match_mode pointer to ha_innobase::m_last_match_mode
@return whether the search mode is unsupported */
template<bool last_match= false>
static bool convert_search_mode_to_innobase(ha_rkey_function find_flag,
                                            page_cur_mode_t &mode,
                                            uint *last_match_mode= nullptr)
```
Furthermore, I am going to deduplicate some conditions in `ha_innobase::records_in_range()`.

Reply:
In fact, that refactoring will shorten both `ha_innobase::index_read()` and `ha_innobase::records_in_range()` by some dozens of bytes. The function (with `last_match=true`) would be inlined in the former, and non-inlined in the latter.

Reply:
a6b365af342969507314804ad7c65fdd78afc481 implements this.